### PR TITLE
Mostly configurable canonical cache size

### PR DESCRIPTION
**Pull request by rphmeier**
Account storage caches (within the cached account entries) aren't currently configurable, but the jump tables and account cache itself are through the `--cache-size-state` flag. This will change in a future PR

### CODE REVIEW

**Comment by gavofyork** [comment]
looks reasonable

**Comment by gavofyork** [comment]
lgtm, but haven't dived deeply into logic. will let @arkpar review before merge.

File: ethcore/src/evm/interpreter/shared_cache.rs [hunk 1]
```diff
@@ -21,25 +21,66 @@ use util::sha3::*;
...
+
+		{
+			let mut cur_size = self.cur_size.lock();
+			*cur_size += d.capacity();
+
+			let mut jump_dests = self.jump_destinations.lock();
+			let cap = jump_dests.capacity();
+
+			// grow the cache as necessary; it operates on amount of items
+			// but we're working based on memory usage.
```
Comment (arkpar) [inline ethcore/src/evm/interpreter/shared_cache.rs]: Would be better to create an a type of LRUCache that manages that and reuse it for the state cache as well. E.g `LRUHeapSizeCache` that requires `HeapSizeOf`. Can be done in a future PR though

File: parity/cli/mod.rs [hunk 2]
```diff
@@ -211,6 +211,8 @@ usage! {
 			or |c: &Config| otry!(c.footprint).cache_size_blocks.clone(),
 		flag_cache_size_queue: u32 = 50u32,
 			or |c: &Config| otry!(c.footprint).cache_size_queue.clone(),
+		flag_cache_size_state: u32 = 25u32,
```
Comment (arkpar) [inline parity/cli/mod.rs]: Should be at least 64MB by default

File: ethcore/src/evm/interpreter/shared_cache.rs (hunk 1 above)
Comment (rphmeier) [inline ethcore/src/evm/interpreter/shared_cache.rs]: I might make an upstream PR to implement that. Contributing back to the commu

Comment (rphmeier) [inline ethcore/src/evm/interpreter/shared_cache.rs]: I've filed an issue at https://github.com/contain-rs/lru-cache/issues/38, I intend to implement and incorporate to the follow-up PR where we'll want this in more than one place.

File: parity/cli/mod.rs (hunk 2 above)
Comment (rphmeier) [inline parity/cli/mod.rs]: I don't agree. 64MB is quite large and I find that there are diminishing returns after a certain point. With the default size, I find that the heavy blocks only process twice as slowly as with an unrestricted cache.

A priority should be placed on keeping the footprint light by default. Miners have both the resources and the technical knowledge to increase the cache size using the flag.

**Pull request merged**